configfile: '../config/config.yaml'
print(config)

##### Target rules #####

import pandas as pd
import os

samples = pd.read_csv ( config['Sample_names'] )

# Create arrays of the sample names
samples = samples['Sample_name'].values
print(samples)


rule all:
    input:
        vcf="variants_atlas/GT_SNP.table",
        dicti= config['ref']['dict'],
        fai= config['ref']['fai'],
        GtoA3p=expand ("results/{sample}/3pGtoA_freq.txt", sample = samples),
        #dbtype="humandb/" + config['params']['annotation']['buildver'] + "_" + config['params']['annotation']['dbtype']+".txt",
        dedup=expand('marked_dedup/{sample}.bam', sample = samples),
        pmd=expand("filtered_pmd/{sample}.bam",sample =  samples),
        pmd_stats=expand("pmd_temp/temp{sample}_stats.sam", sample = samples),
        bam_idx=expand ("filtered_pmd/{sample}.bam.bai", sample = samples),
        #valid = expand ("logs/validate_out/{sample}.log", sample = samples),
        flagstat = expand ("bam/{sample}.bam.flagstat", sample = samples)


##### Modules #####
###################
include: "rules/pre_processing.smk"     
include: "rules/mapping.smk"
include: "rules/post_mapping.smk"
#include: "rules/angsd.smk"

## For Atlas
include: "rules/atlas.smk"
include: "rules/select_variants.smk"

## For GATK haplotype caller tool, classic SNP calling on not so damaged DNA:
#include: "rules/variants_calling.smk"
#include: "rules/annotationHuman.smk"    # WARNING! For annotation of SNP using the ANNOVAR premade database on hg38.
#include: "rules/variants_to_table.smk"